### PR TITLE
[Gnosis Safe] Fix xDAI

### PR DESCRIPTION
**Pull request by anxolin**
# Summary

This is an attempt to improve xDAI support for Gnosis Safe.

Basically, this is what happens in xDAI:
image

I thought at first it would be because of https://eips.ethereum.org/EIPS/eip-3607

But is just a gas estimation error. I think its because of EIP-2929. For some reason, this issue is more severe in xDAI and affect all the ethereum transactions (not just the unwrap)

See why I think it's the case: https://dashboard.tenderly.co/tx/xdai/0x03f99bd6df6e2e952c3242a771a30b00b9f9c73009ae7e7323d2f98b688f8161
image

For more context: https://gnosisinc.slack.com/archives/CQB729LLF/p1635439585259700 

## The Solution
It just need more gas. One way to do this, is set a high enough gas limit as a fallback when it fails to calculate. The estimation fails for the same reason.

I tested it also in mainnet (it costed me a fortune to do the test!)

# To Test

1. Follow a full workflow in xDaI

🚨🚨🚨**IMPORTANT**🚨🚨🚨 
Note that Gnosis Safe has more issues in xDAI! Basically, the last signer will be seeing a message with red letters saying that he/she should reject the transaction because it will fails anyway. That's not true! You only need to raise the gas limit editing the transaction. If you set 200000 as the gas limit, it will work.

## Not included
I noticed that the toast message you see when you approve or wrap/unwrap has the wrong hash, so if you follow the link it will tell you that the transaction doesn't exist.

I will address in another PR

### CODE REVIEW

**Comment by anxolin** [comment]
@elena-zh I believe when you tried the build failed and this is likely why it didn't work.

I hope now should be working. 


NOTE for @alfetopito @W3stside , I tried to use the new endpoint, but for now, I will leave this PR with adding a default gas and I could reiterate once I understand better the issues described here https://gnosisinc.slack.com/archives/C025G521XQD/p1635972794042200

File: src/custom/hooks/useApproveCallback/useApproveCallbackMod.ts [hunk 1]
```diff
@@ -10,6 +10,9 @@ import { calculateGasMargin } from 'utils/calculateGasMargin'
 import { useTokenContract } from 'hooks/useContract'
 import { useActiveWeb3React } from 'hooks/web3'
 import { useTokenAllowance } from 'hooks/useTokenAllowance'
+import { ethers } from 'ethers'
+
+const APPROVE_GAS_LIMIT_DEFAULT = ethers.BigNumber.from('150000')
```
Comment (W3stside) [inline src/custom/hooks/useApproveCallback/useApproveCallbackMod.ts:15]: nit: consider leaving a comment for why 150.000 is the default

**Comment by elena-zh** [comment]
Hey @anxolin , I cleared all site data and tried to wrap a token again: I was not able
image

The waiting modal was immediately closed, I did not receive a request to sign a transaction in Safe.

**Review by W3stside — APPROVED** [review]

**Comment by anxolin** [comment]
> Hey @anxolin , I cleared all site data and tried to wrap a token again: I was not able

It must be using the wrong xDAI endpoint. Let me see what's going on

**Comment by alfetopito** [comment]
Not quite working...

I mean, not without intervention.

Approval failed with the default gas estimation (https://blockscout.com/poa/xdai/tx/0x4bfba2b3b6b993d6bd93e9c765dcbddbb468cb83970b62d240f722a8bbe6605e)
There was no feedback to the UI, so the loading was stuck there forever.
Even on second attempt after I manually set gas limit and it was mined (https://blockscout.com/poa/xdai/tx/0x972c26c51a4bd71908abb76c7f92944308f9e21b984c465f78cba91c9a428609), the UI did not realize that and there was no popup nor notification.
After reloading the page the token was approved though.

screenshot_2021-11-04_11-00-43
screenshot_2021-11-04_11-05-52
screenshot_2021-11-04_11-06-48


Swapping failed for the same reason, gas estimation.

**Comment by alfetopito** [comment]
For context, the Safe I tested has only 1 owner.

Again, it did work after manually adjusting the gas, but it did not recognize the approval.
Is that something to be handled in another PR?

**Comment by anxolin** [comment]
I found the issue with WRAP and fixed it. I hope!

I repetrad the testes for all operations in EOA and Gnosis Safe. All went alright :) 

image

image

File: src/custom/hooks/useApproveCallback/useApproveCallbackMod.ts (hunk 1 above)
Comment (alfetopito) [reply src/custom/hooks/useApproveCallback/useApproveCallbackMod.ts:16]: Hmmm, I did get that error and the default was set 
screenshot_2021-11-04_11-03-25

But the wallet did not receive/did not use this value...

**Review by elena-zh — APPROVED** [review]
Hey @anxolin , I'm able to run transactions in XDAI now ( If I set 200000 as the gas limit). 
When I do not set 200000 as gas limit, all transactions fail and hang in signing state in our app, 
except orders, as orders can be expired.

**Comment by anxolin** [comment]
@alfetopito yes, probably "cancelation/override" of tx won't work for gnosis safe transactions just YET. But this is not related to xDAI

I will add it to the list of items to solve.

Other than that, I'm confused, are your errors related to our app or Gnosis Safe? I mean, have you seen the WARNING in the description?

**Comment by anxolin** [comment]
> Hey @anxolin , I'm able to run transactions in XDAI now ( If I set 200000 as the gas limit).
When I do not set 200000 as gas limit, all transactions fail and hang in signing state in our app,
except orders, as orders can be expired.


This is kind of similar to Leandro comment. Gnosis Safe is also affected by xDAI RPC endpoint failing to estimate, so you need to change the gas or it will fail. We should assume that will be fix on their side and is out of the scope of this PR.

**Comment by anxolin** [comment]
Merged, but let me know leandro if u still think i need to address anything else

**Comment by alfetopito** [comment]
Yeah, got your point from the issue description.
I just thought setting the gas explicitly like you did should be reflected on the final tx.
But I guess that's up to the Safe and the wallet actually triggering the tx.

**Pull request merged**